### ISSUE TITLE
[worklets] <script> tag import feature design.

### ISSUE THREAD

**Issue by bfgeek** [w3c/css-houdini-drafts#376]
From feedback in various forms there was desire to have a declarative way to import scripts initially.

A simple proposal for this would be:
```html
<script type="paintworklet" src="foo.js"></script>
<script type="audioworklet">
registerAudioProcessor(...);
</script>
```

All other attributes would work as expected, e.g. "async", "defer", "nonce", "integrity" have the same behaviour as modules.

(For "async" in particular it'll spin up a global scope, and import the script before the document has finished parsing).

"crossorigin" would be the special one for the moment, which would be forced to "omit" until we decide if we should have cross-origin support.

**Comment by bfgeek** [comment]
cc/ @domenic @annevk thoughts?

**Comment by domenic** [comment]
This seems bad since it will run the code in the main page context if the browser doesn't support worklet="". To avoid that you'd need a new type (e.g. type="paintworklet" or similar).

I think it might also be better to use `<link>` for this, as it's linking to an external resource to be registered somewhere, not running script in the main document. That's what service worker did with `<link rel="serviceworker">`.

**Comment by annevk** [comment]
From OP it seems they want to have inline scripts as well. But yeah, rather than a new worklet attribute the easiest seems to just put it all in type. It's nice that worklets are already JavaScript modules so we don't need that extra dimension here.

**Comment by bfgeek** [comment]
Yes, inline scripts are desirable if html modules get off the ground again, e.g. if this was used inside the [&lt;paper-ripple&gt;](https://github.com/PolymerElements/paper-ripple/blob/master/paper-ripple.html) web component:
```html
<!-- paper-ripple.html web component file -->
<link rel="import" href="../polymer/polymer.html">
<script type="paintworklet">
  // code specific to the paper-ripple web component...
</script>
<template>
  <div id="foo"></div>
</template>
<script>
   // actual web component code....
</script>
``` 

`type="paintworklet"` makes sense, updated the initial comment to reflect that.

**Comment by bfgeek** [comment]
@nyaxt might have thoughts as well.

**Comment by domenic** [comment]
I think it'd be pretty bad to have inline scripts that run in a different global.

**Comment by bfgeek** [comment]
> I think it'd be pretty bad to have inline scripts that run in a different global.

Can you explain your reasoning here?

My concern if we don't allow this is that web developers will simply polyfill this themselves, e.g.

```html
<!-- inside the page or web component file -->
<script id="code" type="paintworklet">
  /* some large complex paint worklet code goes here */
</script>
<script>
  const blob = new Blob([document.getElementById('code').textContent]);
  await CSS.paintWorklet.import(URL.createObjectURL(blob));
</script>
```

Which is bad for performance (duplicates a bunch of memory), etc, hence to provide this natively.

**Comment by annevk** [comment]
@domenic really? We've had the exact same requests for workers in the past. Which also still seems fairly reasonable.

**Comment by annevk** [comment]
(What @bfgeek sketches above is what I've seen for workers as well. It just hasn't surfaced enough for anyone to start caring.)

**Comment by nyaxt** [comment]
From implemetor + spec reader POV, using `<script>` element implies that the worklets are going to be a variation of [script](https://html.spec.whatwg.org/multipage/webappapis.html#concept-script) and scheduled via [script processing model](https://html.spec.whatwg.org/multipage/scripting.html#script-processing-model).
I share the concern w/ @domenic here that having a script variation that doesn't run in the main world environment may be a bit complicated. I'm also not completely sure what {async,defer} would mean for worklet scripts. @domenic's proposal of using `<link>` element is much easier to implement in terms of current Blink.

However, from pure mark-up perspective, I can also be convinced that `<script>` is the most natural way to express any kind of script fragment in the code. Implementing the primitive in Blink isn't technically impossible, but as we are currently actively rationalizing the existing code to actually implement script concept as specified and adding module script concept, it will take some time to deliver.

**Comment by domenic** [comment]
It just seems very strange to me to have scripts like

```html
<script>
self === window;
document.addEventListener(...);
</script>

<script type="module">
self === window;
document.addEventListener(...);
</script>

<script type="paintworklet">
self === window; // fails, window is undefined
document.addEventListener(...); // fails, document is undefined
</script>
```

all in the same source document.

Maybe there is no better option; after all the HTML parser makes it very hard to have any new element contain unstructured data like `<script>` does. But I think it will be surprising for authors and, as @nyaxt expresses, will get in the way of implementations which are so far assuming that `<script>` has a very specific processing model.

In other words, this seems to be a request to reuse the parser behavior of `<script>`, but ignore the rest of its processing model, which is confusing.

If this feature is mostly related to HTML modules, then I'd suggest feeding this back into the discussion there, as it may inform the design there (e.g. using a different HTML parser or different format that isn't HTML as the container).

**Comment by surma** [comment]
I see your point @domenic, and that is indeed weird. From a developer’s perspective all I care about is being able to bundle components into a single file with markup, script, styles, worklets et al. If we can achieve that by feeding that back to HTML modules, then I’m all for that.

**Comment by domenic** [comment]
Let's consolidate this and the other related issues into https://github.com/whatwg/html/issues/6073.
